### PR TITLE
use MINID instead of MAXLEN for better atomicity of the operation

### LINKED ISSUE #296: The worker lost RPC calls when there is a lot of calls made into the method
## Summary
The RPC make uses of the redis stream to store all the pending calls. In myriad this stream can be manipulated by the following futures:
1. The worker to dispatch the RPC call (in the service)
2. Cleanup worker to remove the old item (in the service)
3. Caller (can be either inside or outside the service)

These future are running in parallel, and usually it is safe to use under normal circumstances. 
However inside the high load situation, the cleanup worker (https://github.com/deriv-com/perl-Myriad/blob/master/lib/Myriad/Transport/Redis.pm#L298-L336) might remove too much item than it supposed to be due to the assumption that stream length would be constant during the stream length computation (which introduces the race window). 

To reproduce this, the calling rate need to be higher than the rate the calls are picked up.

## Possible fixes
- Redis 6.2.0 added a new flag which allow us to trim the stream so it would retains all item after the specified ID
- To reduce the race window by considering the number of the entries added between the stream size calculation

Attached this patch as workaround for now:
```diff
diff --git a/lib/Myriad/Transport/Redis.pm b/lib/Myriad/Transport/Redis.pm
index 97e0b15..bd73f13 100644
--- a/lib/Myriad/Transport/Redis.pm
+++ b/lib/Myriad/Transport/Redis.pm
@@ -329,8 +329,16 @@ async method cleanup (%args) {
         }
         $total = $info->{length} - $total if $direction eq 'xrange';
 
+        # As the cleanup operation is NOT atomic, this will introduce a race window
+        # when the caller makes calls in a high rate, we should compensate the number of entries
+        # added during the calculation
+        my ($info2) = await $self->stream_info($stream);
+        my $entries_added = ($info2->{length} - $info->{length});
+        # Add the entries added during the process
+        # gaurd against race trimming by multiple instances of the service
+        $total += int($entries_added);
 #        my ($before) = await $redis->memory_usage($stream);
-        my ($trim) = await $redis->xtrim($stream, MAXLEN => $total);
+        my ($trim) = await $redis->xtrim($stream, MAXLEN => $total) unless $entries_added < 0;
 #        my ($after) = await $redis->memory_usage($stream);
         $log->tracef('Trimmed %d items from stream: %s', $total, $stream);
     }
```

### PR DESCRIPTION
**Pull request by ngeojiajun-deriv**
Fixes #296

### CODE REVIEW

File: lib/Myriad/Transport/Redis.pm [hunk 1]
```diff
@@ -296,42 +296,7 @@ async method cleanup (%args) {
...
-            # Overlapping ranges, so the next ID will be included twice
-            --$total;
-            $endpoint = $v->[-1][0];
-        }
-        $total = $info->{length} - $total if $direction eq 'xrange';
-
-#        my ($before) = await $redis->memory_usage($stream);
-        my ($trim) = await $redis->xtrim($stream, MAXLEN => $total);
-#        my ($after) = await $redis->memory_usage($stream);
+        my ($total) = await $redis->xtrim($stream, MINID => $oldest);
```
Comment (tom-binary) [inline lib/Myriad/Transport/Redis.pm:299]: I'd suggest using approximate trimming, with a config option to allow exact trimming if needed.

```suggestion
        my ($total) = await $redis->xtrim($stream, MINID => ($exact_trim ? () : '~'), $oldest);
```

Comment (ngeojiajun-deriv) [reply lib/Myriad/Transport/Redis.pm:299]: Should we add a environment variable option `MYRIAD_REDIS_TRIM_EXACT` for this purpose?

Comment (nael-binary) [reply lib/Myriad/Transport/Redis.pm:299]: yeah I think worth doing that.
we would need to add `trim_exact` in transport class (`MYRIAD_REDIS_TRIM_EXACT`) in config class
along with Myriad::Config class (can be in our %DEFAULTS)
while default is approximate trimming like Tom mentioned.

**Review by diego-deriv — APPROVED** [review]

File: lib/Myriad/Config.pm [hunk 2]
```diff
@@ -50,24 +50,25 @@ alternative.
...
-    storage_transport      => undef,
-    transport              => 'redis',
-    metrics_adapter        => 'Statsd',
-    metrics_host           => 'localhost',
-    metrics_port           => '8125',
+    services                        => { },
+    config_path                     => 'config.yml',
+    transport_redis                 => 'redis://localhost:6379',
+    transport_redis_cache           => 0,
+    transport_redis_exact_trimming  => 0,
```
Comment (nael-binary) [inline lib/Myriad/Config.pm:57]:
```suggestion
    transport_redis_trim_exact  => 0,
```

File: lib/Myriad/Transport/Redis.pm [hunk 3]
```diff
@@ -65,6 +65,7 @@ declare_exception 'NoSuchStream' => (
 );
 
 field $use_cluster;
+field $use_exact_trimming
```
Comment (nael-binary) [inline lib/Myriad/Transport/Redis.pm:68]:
```suggestion
field $use_trim_exact
```

File: lib/Myriad/Transport/Redis.pm [hunk 4]
```diff
@@ -296,42 +298,7 @@ async method cleanup (%args) {
...
-            # Overlapping ranges, so the next ID will be included twice
-            --$total;
-            $endpoint = $v->[-1][0];
-        }
-        $total = $info->{length} - $total if $direction eq 'xrange';
-
-#        my ($before) = await $redis->memory_usage($stream);
-        my ($trim) = await $redis->xtrim($stream, MAXLEN => $total);
-#        my ($after) = await $redis->memory_usage($stream);
+        my ($total) = await $redis->xtrim($stream, MINID => ($use_exact_trimming ? () : '~'), $oldest);
```
[the remark this answers is not in the document]
Comment (ngeojiajun-deriv) [reply lib/Myriad/Transport/Redis.pm:301]: the same command attempted in production, it is working as expected

File: lib/Myriad/Transport/Redis.pm (hunk 3 above)
Comment (tom-binary) [reply lib/Myriad/Transport/Redis.pm:68]: also needs the `;`

**Review by nael-binary — APPROVED** [review]

**Pull request merged**